pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+	// 	klog.Errorf("Reconciler Failed to list ops: %v", err)
+	// }
+	instances, shares, instanceShares := recon.managedInstanceAndShare(instances, shares)
+
+	shareInfoMap := recon.createAndUpdateShareInfos(shares)
+
+	shareInfoList, err := recon.shareLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list ShareInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("listed out %d shareInfo objects", len(shareInfoList))
+	for _, shareInfo := range shareInfoList {
+		shareName := shareInfo.Name
+		if _, ok := shareInfoMap[shareName]; !ok {
+			shareInfoMap[shareName] = shareInfo
+		}
+	}
+
+	instanceInfoMap := recon.createAndUpdateInstanceInfos(instances, instanceShares)
+
+	instanceInfoList, err := recon.instanceLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list InstanceInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("listed out %d instanceInfo objects", len(instanceInfoList))
+	for _, instanceInfo := range instanceInfoList {
+		instanceURI := util.IINameToInstanceURI(instanceInfo.Name)
+		if _, ok := instanceInfoMap[instanceURI]; !ok {
+			instanceInfo, err := recon.maybeRemoveInstanceInfoFinalizer(instanceInfo)
+			if err != nil {
+				klog.Errorf("error removing Finalizer from InstanceInfo %q: %v", instanceInfo.Name, err)
+				continue
+			}
+			if instanceInfo != nil {
+				instanceInfoMap[instanceURI] = instanceInfo
+			}
+		}
+	}
+
+	recon.assignSharesToInstances(shareInfoMap, instanceInfoMap)
+
+	klog.Infof("reconciliation round finished after %v", time.Since(startTime))
+}
+
+func (recon *multishareReconciler) assignSharesToInstances(shareInfos map[string]*v1alpha1.ShareInfo, instanceInfos map[string]*v1alpha1.InstanceInfo) {
+	for _, instanceInfo := range instanceInfos {
+		if instanceInfo.Status == nil {
+			klog.Infof("instance %q has Status nil", instanceInfo.Name)
+			continue
+		}
+
+		instanceInfoClone := instanceInfo.DeepCopy()
+		updated := false
+
+		instanceURI := util.IINameToInstanceURI(instanceInfo.Name)
+		for _, shareName := range instanceInfo.Status.ShareNames {
+			shareInfo, ok := shareInfos[shareName]
+			if !ok {
+				klog.Errorf("share %q is assigned to instance %q but shareInfo does not exist", shareName, instanceURI)
+				continue
+			}
+			if shareInfo.Status == nil || shareInfo.Status.InstanceHandle == "" {
+				shareInfo, err := recon.assignShareInfo(shareInfo, instanceURI)
+				if err != nil {
+					klog.Errorf("cannot update shareInfo %q's instanceHandle: %v", shareName, err)
+					continue
+				}
+				shareInfos[shareName] = shareInfo
+			} else if shareInfo.Status.InstanceHandle != instanceURI {
+				klog.Warningf("instanceInfo %q has share %q but its shareInfo points to instance %q", instanceInfo.Name, shareName, shareInfo.Status.InstanceHandle)
+
+				// if the share is already marked for deletion, don't try to add it to the instanceInfo it points to.
+				if shareInfo.DeletionTimestamp == nil {
```

Comment:
can we flip the check to avoid indentation?

if shareInfo.Deletetimestamp != nil {
   // remove share from instance info
}

// fix the instanceInfo

Reply:
the "remove share form instanceInfo" step needs to happen either way. So I think flipping the check makes it harder to read.

Reply:
Ack